src/pybind/mgr/call_home_agent/module.py

```diff
@@ -0,0 +1,322 @@
+    "events": []
+}
+
+# TODO: report functions providing the json paylod should be imported from the data collectors module
+
+
+def report_inventory() -> str:
+    """
+    Produce the inventory report
+    """
+    dt = datetime.timestamp(datetime.now())
+    # Set timestamps
+    reports_header['event_time'] = datetime.fromtimestamp(dt).strftime("%Y-%m-%d %H:%M:%S")
+    reports_header['event_time_ms'] = str(int(dt))
+    reports_header['local_event_time'] = datetime.fromtimestamp(dt).strftime("%a %b %d %H:%M:%S %Z")
+
+    # Set event id
+    reports_header['event_id'] = "IBM_chc_event_RedHatMarine_ceph_{}_daily_report_{}".format(
+                                            reports_header['asset_virtual_id'],
+                                            reports_header['event_time_ms'])
+
+    # return json.dumps({**reports_header, **inventory()}) <---- TODO ADD INVENTORY (from datacollector) as event
+    return json.dumps(reports_header)
+
+
+def report_performance() -> str:
+    """
+    TODO: temporally uses inventory report
+    """
+    return report_inventory()
+
+
+class Report:
+    def __init__(self, name: str, fn: Callable[[], str], url: str, minutes_interval: int):
+        self.name = name                       # name of the report
+        self.fn = fn                           # function used to retrieve the data
+        self.url = url                         # url to send the report
+        self.interval = minutes_interval * 60  # interval to send the report (seconds)
+
+    def print(self) -> str:
+        if reports_header['private_key'] and reports_header['api_key']:
+            return self.fn()
+        else:
+            raise Exception('Not able to print <%s> report. Identification keys are not available' % self.name)
+
+    def send(self) -> str:
+        fail_reason = ""
+        resp = None
+        if reports_header['private_key'] and reports_header['api_key']:
+            try:
+                resp = requests.post(url=self.url,
+                                     headers={'accept': 'application/json', 'content-type': 'application/json'},
+                                     data=self.fn())
+                resp.raise_for_status()
+            except Exception as e:
+                explanation = resp.content if resp else ''
+                fail_reason = 'Failed to send <%s> to <%s>: %s \n%s' % (self.name, self.url, str(e), explanation)
+        else:
+            fail_reason = 'Not able to send <%s> report. Identification keys are not available' % self.name
+        return fail_reason
+
+
+class CallHomeAgent(MgrModule):
+    MODULE_OPTIONS: List[Option] = [
+        Option(
+            name='ches_endpoint',
+            type='str',
+            default=CHES_ENDPOINT,
+            desc='Call Home Event streamer end point'
+        ),
+        Option(
+            name='interval_inventory_report_minutes',
+            type='int',
+            default=INTERVAL_INVENTORY_REPORT_MINUTES,
```

Comment:
Both interval settings are a bit tricky, because they're based on a wait(), but what happens if someone restarts the mgr module? The report will be immediately sent. And if they restart it multiple times (which is not that unusual), multiple reports will be sent either.

The only way to deal with is is:
* To persist in the mon KV (set_store/get_store) the timestamp of the last report and re-send as soon as that timestamp is older than the interval.
* Use a stateless approach: simplifying, if current time is 00:00, then send report. This still needs some refinement (something like `(now() / wait_time) % (interval / wait_time)`).
* Scheduled runs: rather than keeping a mostly idle loop, the mgr module is just one-shot command-based (no `serve` loop method implemented), and from `crontab` users can periodically invoke the `ceph callhome send` command. 

BTW, one issue still remains: if the cluster is down, or at least mgr's are, no reports are sent. This means that we might be missing some interesting clusters.

Reply:
I see two ways forward without making the implementation too complex:

1. Set an interval that is prepended to the action. Meaning that we would wait 5 minutes after a cluster restart until we actually send data. That way we would avoid re-sending over and over at the cost of maybe sending nothing until the cluster stabilizes.
2. Implement the field to use cron time format. This would make it harder for the user to set it correctly, but we would be more flexible than with a mere "minutes" field.
For "every 5 minutes" we would set the field to `*/5 * * * *` for example